File: src/common/Order.cs
using LanguageExt;
using LanguageExt.Common;
using System.Collections.Immutable;

namespace common;

public sealed record OrderId
{
    private OrderId(string value) => Value = value;

    public string Value { get; }

    public static Fin<OrderId> From(string value) =>
        string.IsNullOrWhiteSpace(value)
            ? Error.New("Order ID cannot be null or whitespace.")
            : new OrderId(value);

    public static OrderId FromOrThrow(string value) =>
        From(value).ThrowIfFail();
}

public sealed record Order
{
    public required OrderId Id { get; init; }
    public required ImmutableArray<Pizza> Pizzas { get; init; }
}

File: src/common/Order.fs
namespace common

open System

type OrderId =
    private
    | OrderId of string

    static member fromString value =
        if String.IsNullOrWhiteSpace value then
            invalidOp "Order ID cannot be empty."
        else
            OrderId value

    static member toString(OrderId value) = value

type OrderStatus =
    | Pending
    | Canceled

    static member toString value =
        match value with
        | Pending -> nameof Pending
        | Canceled -> nameof Canceled

    static member fromString value =
        match value with
        | _ when String.Equals(value, nameof Pending, StringComparison.OrdinalIgnoreCase) -> Pending
        | _ when String.Equals(value, nameof Canceled, StringComparison.OrdinalIgnoreCase) -> Canceled
        | _ -> invalidOp ($"'{value}' is not a valid order status.")

type Order =
    { Id: OrderId
      Status: OrderStatus
      Pizzas: Pizza list }
